src/websocket.c

```diff
@@ -77,10 +77,26 @@ ws_compute_handshake(struct http_client *c, char *out, size_t *out_sz) {
 	return 0;
 }
 
+
+void
+ws_write_all(int socket, void *buffer, size_t length) {
+    size_t written = 0;
+    int n;
+    while (written < length) {
+        int n = write(socket, (char*)buffer + written, length - written);
+        if (n <= 0) {
+            if (errno == EINTR || errno == EAGAIN)
+                continue;
+            err(EXIT_FAILURE, "Could not write() data");
```

Comment:
This should use the logging interface instead of terminating the service with `err`. You can pass a pointer to the `server` struct from the caller and use it like this:

```c
#include "slog.h"

// ...

int
ws_write_all(struct server *s, int fd, void *buffer, size_t length) {

// ...

const char err_msg[] = "Could not write data to websocket";
slog(s, WEBDIS_ERROR, err_msg, sizeof(err_msg)-1);
return -1;
```

Instead of calling `err`, a return code should be used to tell the caller about the failure so that it can cleanly terminate the connection and free the WS structure.
Successful writes should `return 0;` after the `while` loop. See comment below too.

Unfortunately I see that even if this error is propagated from `ws_write_all` to `ws_handshake_reply` to the caller of `ws_handshake_reply`, the return code is unused in `worker.c`. There, it should instead be used to set `c->broken = 1` just like a couple of lines below the call to `ws_handshake_reply`.

Something like this. Instead of:
```c
if(nparsed < ret) {
	http_client_add_to_body(c, c->buffer + nparsed + 1, c->sz - nparsed - 1);
	ws_handshake_reply(c);
}
```

you'd want to use the return value:
```c
if(nparsed < ret) {
	http_client_add_to_body(c, c->buffer + nparsed + 1, c->sz - nparsed - 1);
	if(ws_handshake_reply(c) < 0)
		c->broken = 1;
}
```
I'm also not sure about this repeated write outside of the event loop. Ideally the call to `write(2)` should happen when the file descriptor is writable, and the event re-registered if there's still more data to write. This is a more complex change though, and could be done in a second pass.